src/NServiceBus.Core/Transports/DeliveryConstraintCollection.cs

```diff
@@ -0,0 +1,44 @@
+namespace NServiceBus.Transports
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using NServiceBus.DeliveryConstraints;
+
+    /// <summary>
+    /// A collection of delivery constraints.
+    /// </summary>
+    public class DeliveryConstraintCollection
+    {
+        List<DeliveryConstraint> constraints;
+
+        internal DeliveryConstraintCollection(IEnumerable<DeliveryConstraint> constraints)
+        {
+            this.constraints = constraints.ToList();
+        }
+
+        /// <summary>
+        /// Tries to get and remove constraint of a given type.
+        /// </summary>
+        /// <typeparam name="T">Requested type.</typeparam>
+        /// <param name="constraint">Returned constraint or null.</param>
+        /// <returns>Success or failure</returns>
+        public bool TryRemove<T>(out T constraint) where T : DeliveryConstraint
```

Comment:
I am fine with `Maybe` if we all agree this is the way to go. While I don't have any strong opinions, my weak opinion is use whatever idioms the language provides. Since C# is not a functional language, I would default to `out`.